components/container_id/container_id.c

```diff
@@ -1,101 +1,113 @@
-    // Example Fargate
-    // 11:something:/ecs/5a081c13-b8cf-4801-b427-f4601742204d/432624d2150b349fe35ba397284dea788c2bf66b885d14dfc1569b01890ca7da
-    // Example Fargate 1.4+
-    // 1:name=systemd:/ecs/34dc0b5e626f2c5c4c5170e34b10e765-1234567890
-    {":/ecs/", sizeof(":/ecs/") - 1},
-};
-static size_t dd_targets_len = sizeof dd_targets / sizeof dd_targets[0];
-
-/* Updates the 'end' pointer to the last non-whitespace & non-null-terminating
- * character of a 'str'
+#define LINE_REGEX "^[0-9]+:[^:]*:.+$"
+/* Examples of some valid container IDs.
+ *
+ * Docker
+ * 13:name=systemd:/docker/3726184226f5d3147c25fdeab5b60097e378e8a720503a5e19ecfdf29f869860
+ *
+ * Kubernetes
+ * 11:perf_event:/kubepods/something/pod3d274242-8ee0-11e9-a8a6-1e68d864ef1a/3e74d3fd9db4c9dd921ae05c2502fb984d0cde1b36e581b13f79c639da4518a1
+ *
+ * ECS
+ * 9:perf_event:/ecs/user-ecs-classic/5a0d5ceddf6c44c1928d367a815d890f/38fac3e99302b3622be089dd41e7ccf38aff368a86cc339972075136ee2710ce
+ *
+ * Fargate < 1.4
+ * 11:something:/ecs/5a081c13-b8cf-4801-b427-f4601742204d/432624d2150b349fe35ba397284dea788c2bf66b885d14dfc1569b01890ca7da
  */
-static size_t dd_str_seek_end(char *str, char **end) {
-    size_t len = strlen(str);
-    *end = str + len - 1;
-    while (*end > str && isspace(*end[0])) {
-        --*end;
-        --len;
-    }
-    return len;
-}
+#define CONTAINER_REGEX "[0-9a-f]{64}"
+/* Example of a valid task ID.
+ *
+ * Fargate 1.4+
+ * 1:name=systemd:/ecs/34dc0b5e626f2c5c4c5170e34b10e765-1234567890
+ */
+#define TASK_REGEX "[0-9a-f]{32}-[0-9]+"
 
-static void dd_extract_id(char *buf, char *pos) {
-    char *end = NULL;
-    size_t len = dd_str_seek_end(pos, &end);
+#define MIN_ID_LEN DATADOG_PHP_CONTAINER_ID_MIN_LEN
+#define MAX_ID_LEN DATADOG_PHP_CONTAINER_ID_MAX_LEN
 
-    if (len < DATADOG_PHP_CONTAINER_ID_MAX_LEN) {
-        return;
-    }
+static size_t dd_find_id_regex(regex_t *regex, char *buf, const char *line) {
+    size_t len;
+    regmatch_t pmatch[1];
 
-    // TODO Refactor to support Fargate 1.4+ IDs
-    // [a-f0-9]{64}
-    for (size_t i = DATADOG_PHP_CONTAINER_ID_MAX_LEN; i > 0; i--) {
-        char c = end[0];
-        if (!isxdigit(c)) {
-            buf[0] = '\0';
-            return;
-        }
-        buf[i - 1] = c;
-        end--;
-    }
-    buf[DATADOG_PHP_CONTAINER_ID_MAX_LEN] = '\0';
-}
+    if (regexec(regex, line, 1, pmatch, 0) != 0) return 0;
 
-static char *dd_find_target(const char *line) {
-    for (size_t i = 0; i < dd_targets_len; ++i) {
-        char *pos = strstr(line, dd_targets[i].name);
-        if (pos) {
-            return pos + dd_targets[i].len;
-        }
-    }
-    return NULL;
+    len = (size_t)pmatch[0].rm_eo - pmatch[0].rm_so;
+    if (len < MIN_ID_LEN || len > MAX_ID_LEN) return 0;
+
+    memcpy(buf, line + pmatch[0].rm_so, len);
+    buf[len] = '\0';
+
+    return len;
 }
 
 void datadog_php_container_id(char *buf, const char *file) {
```

Comment:
If this "fails", how does the caller know?

Reply:
Given a valid `buf` is provided by the caller, it will always be filled with a container ID or an empty string. But the caller won't know if the empty string means there was an error leading up to parsing the cgroup file or if there was no valid container ID found in the file.

Do you think returning an error code would be better so that the caller could choose to emit a debug message if there was an error parsing the cgroup file?